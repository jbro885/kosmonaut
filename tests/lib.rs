use crate::util::CommandUnderTest;
use std::ffi::OsStr;

mod util;

mod layout;
mod style;

pub fn reftest_expect_similar<S: AsRef<str> + AsRef<OsStr>>(file_path_one: S, file_path_two: S) {
    let (file_path_one, file_path_two): (&str, &str) =
        (file_path_one.as_ref(), file_path_two.as_ref());
    match run_similarity_cmd(file_path_one, file_path_two) {
        Ok(percent_similar) => {
            if percent_similar != 100.0f64 {
                panic!(format!("reftest_expect_similar failure.  files '{}' and '{}' were only {}% similar (expected 100%)", file_path_one, file_path_two, percent_similar));
            }
        }
        Err(err) => panic!(format!(
            "reftest_expect_similar *internal* failure (this is bad) with files '{}' and '{}'.  error message was: {}",
            file_path_one,
            file_path_two,
            err
        )),
    }
}

fn run_similarity_cmd<S: AsRef<str> + AsRef<OsStr>>(
    file_path_one: S,
    file_path_two: S,
) -> Result<f64, String> {
    let mut cmd = CommandUnderTest::new();
    cmd.arg("similarity");
    cmd.arg("--files");
    cmd.arg(file_path_one);
    cmd.arg(file_path_two);
    cmd.arg("--similarity-percent-only");
    cmd.succeeds();
    let stdout = cmd.stdout().trim();
    stdout.parse::<f64>().map_err(|_| stdout.trim().to_owned())
}
