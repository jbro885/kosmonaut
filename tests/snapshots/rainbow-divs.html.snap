---
source: tests/lib.rs
expression: dump_layout_cmd.stdout()
input_file: tests/websrc/rainbow-divs.html
---
HTML BlockContainer at (0, 0) size 1920x184
  BODY BlockContainer at (8, 8) size 1904x168
    DIV BlockContainer at (20, 20) size 1880x144
      DIV BlockContainer at (32, 32) size 1856x120
        DIV BlockContainer at (44, 44) size 1832x96
          DIV BlockContainer at (56, 56) size 1808x72
            DIV BlockContainer at (68, 68) size 1784x48
              DIV BlockContainer at (80, 80) size 1760x24
                DIV BlockContainer at (92, 92) size 1736x0

